### PR TITLE
Add file backed state to cpu manager

### PR DESCRIPTION
**Pull request by flyingcougar**
**What this PR does / why we need it**:
Adds file backed `State` implementation to cpu manger with tests.
Reads from `State` are done from memory, while each write triggers state save to a file.

Any failure in reading the state file results in empty state

Next PR: #54409

### CODE REVIEW

**Comment by flyingcougar** [comment]
ping @derekwaynecarr @balajismaniam @ConnorDoyle

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_file.go]: Nit: `s/parse/serialize`
since we are writing, not reading here

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_file.go]: `s/created/written`

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_file.go]: Could we return an `error` from this function? Then I think you could clean up a lot of the redundant condition checks by bailing as soon as you fail to init the file. At the call site, if err is not nil then you can just call `storeState()`. We may want another check in `NewFileState` after all that to simply `stat` the file and panic if it still doesn't exist.

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_file.go]: Could we make the JSON key and the struct key match? e.g. make the json key simply `entries`

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_file.go]: Nit: another (non-verb) name for this would be preferable. How about `data`?

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_file.go]: None of the callers use the return value. Suggest to either drop the return value or use it in the callers.

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_file.go]: Why does this only log, in comparison to the write failure below where we panic? The end result is the same (fail to write the file.)

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_file.go]: Let's make this a regular member instead of embedding the interface. That way if we add functions to the `state.State` interface later, the compiler will help us remember to implement them for this type instead of inheriting implementations from the memory state. Maybe call the member `cache`?

File: pkg/kubelet/cm/cpumanager/state/state_file_test.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_file_test.go]: last arg should be `cpuassignmentsSm`?

File: pkg/kubelet/cm/cpumanager/state/state_file_test.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_file_test.go]: Option: use `%s` to get the canonical string out of the cpuset. Your choice though, whatever output is more helpful.

File: pkg/kubelet/cm/cpumanager/state/state_file_test.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_file_test.go]: `!cpusetSf.Equals(cpusetSm)`

File: pkg/kubelet/cm/cpumanager/state/state_file_test.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_file_test.go]: Cool!

File: pkg/kubelet/cm/cpumanager/state/state_file_test.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_file_test.go]: let's use https://golang.org/pkg/io/ioutil/#TempFile instead

File: pkg/kubelet/cm/cpumanager/state/state_file_test.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_file_test.go]: let's use https://golang.org/pkg/io/ioutil/#TempFile instead

**Review by ConnorDoyle — CHANGES_REQUESTED** [review]
Really nice work; please take a look at the comments. Once fixed, let's land it!

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/state/state_file.go]: :+1:

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/state/state_file.go]: ok, can do, i just wanted to avoid implementing all `Get` methods

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/state/state_file.go]: :+1:

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/state/state_file.go]: :+1

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/state/state_file.go]: :+1:

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/state/state_file.go]: :+1:

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/state/state_file.go]: :+1:

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/state/state_file.go]: :+1:

**Comment by ConnorDoyle** [comment]
Flaked due to TestDevicePluginReRegistration (flake issue: #52818)

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_file.go]: Unnecessary write here? Looks like ClearState calls storeState already.

**Comment by ConnorDoyle** [comment]
/approve

Will add LGTM when tests become green again. Thanks @flyingcougar

File: pkg/kubelet/cm/cpumanager/state/state_file.go
Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/state/state_file.go]: this is `cache.ClearState()` not `stateFile.ClearState()`

**Pull request merged**
